### ISSUE TITLE
Stuck in 'connect' state with Cluster and Node 10.x

### ISSUE THREAD

**Issue by brettkiefer** [luin/ioredis#633]
I'm seeing ioredis 3.2.2 intermittently getting indefinitely stuck in the 'connect' state when run with Redis.Cluster under all versions of Node 10 up to the latest (10.3.0), and not under previous major versions (e.g. 9.11.1). I'm working on a minimal repro to illustrate this, but right now it looks as though the Node 10 stream is not emitting a 'close' event to trigger `eventHandler.closeHandler(_this)`  under Node 10.

I'm not yet sure why this happens, but I thought I'd give a heads-up as I look at it in case anyone else is seeing the same.

**Comment by brettkiefer** [comment]
It's slightly harder to get it to stall out with a single connection (dialing down nRedis to 1), but after a few runs of that on Node 10.3.0 and IORedis debug enabled, I get:

```
Thu, 31 May 2018 20:46:52 GMT ioredis:cluster status: [empty] -> connecting
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis1:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis2:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis3:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis4:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis5:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis6:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:52 GMT ioredis:cluster getting slot cache from redis5:6379
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis5:6379]: wait -> connecting
Thu, 31 May 2018 20:46:52 GMT ioredis:redis queue command[0] -> cluster(slots)
Thu, 31 May 2018 20:46:52 GMT ioredis:redis status[redis2:6379]: wait -> connecting
Start burning CPU
Stop burning CPU, burned 3693ms
Thu, 31 May 2018 20:46:56 GMT ioredis:cluster getting slot cache from redis2:6379
Thu, 31 May 2018 20:46:56 GMT ioredis:redis queue command[0] -> cluster(slots)
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:46:56 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:46:56 GMT ioredis:redis queue command[0] -> info()
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Thu, 31 May 2018 20:46:56 GMT ioredis:connection send 1 commands in offline queue
Thu, 31 May 2018 20:46:56 GMT ioredis:redis write command[0] -> cluster(slots)
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis1:6379]: wait -> close
Thu, 31 May 2018 20:46:56 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis1:6379]: close -> end
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis3:6379]: wait -> close
Thu, 31 May 2018 20:46:56 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis3:6379]: close -> end
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis4:6379]: wait -> close
Thu, 31 May 2018 20:46:56 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis4:6379]: close -> end
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis6:6379]: wait -> close
Thu, 31 May 2018 20:46:56 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[redis6:6379]: close -> end
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:46:56 GMT ioredis:cluster status: connecting -> connect
Thu, 31 May 2018 20:46:56 GMT ioredis:redis queue command[0] -> cluster(info)
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: ready -> close
Thu, 31 May 2018 20:46:56 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: close -> end
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: wait -> connecting
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:46:56 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:46:56 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Set redis key on cluster 0
### STUCK: CLUSTER (0) STALLED OUT ###
```

**Comment by brettkiefer** [comment]
On a healthy run (this is with Node 10.3.0 as well, but just on a run where we don't see the intermittent failure), we see the cluster get into the ready state instead of stalling out in 'connect':

```
Thu, 31 May 2018 20:56:09 GMT ioredis:cluster status: [empty] -> connecting
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis1:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis2:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis3:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis4:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis5:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis6:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:09 GMT ioredis:cluster getting slot cache from redis6:6379
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis6:6379]: wait -> connecting
Thu, 31 May 2018 20:56:09 GMT ioredis:redis queue command[0] -> cluster(slots)
Thu, 31 May 2018 20:56:09 GMT ioredis:redis status[redis2:6379]: wait -> connecting
Start burning CPU
Stop burning CPU, burned 3359ms
Thu, 31 May 2018 20:56:12 GMT ioredis:cluster getting slot cache from redis1:6379
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis1:6379]: wait -> connecting
Thu, 31 May 2018 20:56:12 GMT ioredis:redis queue command[0] -> cluster(slots)
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis queue command[0] -> info()
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Thu, 31 May 2018 20:56:12 GMT ioredis:connection send 1 commands in offline queue
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> cluster(slots)
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis3:6379]: wait -> close
Thu, 31 May 2018 20:56:12 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis3:6379]: close -> end
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis4:6379]: wait -> close
Thu, 31 May 2018 20:56:12 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis4:6379]: close -> end
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis5:6379]: wait -> close
Thu, 31 May 2018 20:56:12 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[redis5:6379]: close -> end
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: [empty] -> wait
Thu, 31 May 2018 20:56:12 GMT ioredis:cluster status: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: ready -> close
Thu, 31 May 2018 20:56:12 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: close -> end
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: ready -> close
Thu, 31 May 2018 20:56:12 GMT ioredis:connection skip reconnecting since the connection is manually closed.
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: close -> end
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: wait -> connecting
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Thu, 31 May 2018 20:56:12 GMT ioredis:cluster getting slot cache from xx.xxx.x.xx:6379
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> cluster(slots)
Thu, 31 May 2018 20:56:12 GMT ioredis:delayqueue send 1 commands in failover queue
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: wait -> connecting
Thu, 31 May 2018 20:56:12 GMT ioredis:redis queue command[0] -> cluster(info)
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connecting -> connect
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> info()
Thu, 31 May 2018 20:56:12 GMT ioredis:redis status[xx.xxx.x.xx:6379]: connect -> ready
Thu, 31 May 2018 20:56:12 GMT ioredis:connection send 1 commands in offline queue
Thu, 31 May 2018 20:56:12 GMT ioredis:redis write command[0] -> cluster(info)
Thu, 31 May 2018 20:56:12 GMT ioredis:cluster status: connect -> ready
Set redis key on cluster 0
Thu, 31 May 2018 20:56:13 GMT ioredis:redis write command[0] -> set(test.set.key,1527800173790)
Set redis key on cluster 0
Thu, 31 May 2018 20:56:14 GMT ioredis:redis write command[0] -> set(test.set.key,1527800174791)
Set redis key on cluster 0
Thu, 31 May 2018 20:56:15 GMT ioredis:redis write command[0] -> set(test.set.key,1527800175792)
<<< SUCCESS, WE DID NOT STALL OUT >>>
```

**Comment by brettkiefer** [comment]
This is the most minimal simulation I have been able to generate for what we're seeing for the good and bad cases in our actual application in staging on Node 10 and Cluster 3.2.2. This may boil down to a Node bug or some kind of improper usage on our end, but it seems like understanding what's causing the cluster to get stuck in 'connect' as the cluster(info) command goes through and the redis goes to 'ready' is the beginning, and that's where I'm having some trouble.

**Comment by brettkiefer** [comment]
Unsurprisingly (this was the most suspicious-looking change in the (changelog)[https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md] for this), this traces back to https://github.com/nodejs/node/commit/9b7a6914a7. If you build Node at the previous revision (https://github.com/nodejs/node/commit/74553465e6), this test does not repro, at 9b7a6914a7f0bd754e78b42b48c75851cfd6b3c4 it does.

It looks like there is some hubbub around 'close' events in the subsequent Node 10 changelogs, but concerning http, e.g. https://github.com/nodejs/node/pull/20941.

**Comment by ccs018** [comment]
So you're suspecting that this is really a node issue and not an ioredis issue?  Would probably to be good to log an issue there as well.  Try to get some of those folks to re-review that commit.  This issue is going to prevent folks using ioredis to upgrade to Node 10.

**Comment by brettkiefer** [comment]
Yes, I'm suspecting that Node isn't emitting 'close' under certain circumstances, and that that is a bug with that commit, but it'll probably take me a little while to boil it down to something that takes ioredis out of it.

Currently we can work around this for the purposes of this test by handling 'end' rather than close

diff --git a/lib/redis.js b/lib/redis.js
index 8870bec..39b9b53 100644
--- a/lib/redis.js
+++ b/lib/redis.js
@@ -279,7 +279,7 @@ Redis.prototype.connect = function (callback) {

       stream.once(CONNECT_EVENT, eventHandler.connectHandler(_this));
       stream.once('error', eventHandler.errorHandler(_this));
-      stream.once('close', eventHandler.closeHandler(_this));
+      stream.once('end', eventHandler.closeHandler(_this));
       stream.on('data', eventHandler.dataHandler(_this));

... but the 'close' handling seems more correct to me, and it looks as though that event should always be emitted.

**Comment by brettkiefer** [comment]
Ah, ok, if the socket is paused there may not be a 'close':

https://github.com/nodejs/node/pull/19241#issuecomment-371639579

**Comment by brettkiefer** [comment]
... but at the point when we get the EOF (https://github.com/nodejs/node/commit/9b7a6914a7#diff-e6ef024c3775d787c38487a6309e491dL645), the connection isPaused() returns false, so it's not that.

The issue is that we get into onReadableStreamEnd with the socket's this.writable set to true, so the socket's maybeDestroy doesn't call socket.destroy and the socket never emits closed.

It's a little hard to tell if that's intentional or not, from the PR and checkin, but it's looking more like 'not'.

**Comment by brettkiefer** [comment]
There is more on what exactly it looks like is happening with the socket at https://github.com/nodejs/node/pull/19241#issuecomment-396133420 -- the long and short of it appears to be that if we hit the connect timeout and call .end before the socket is connected, the socket gets in a state that causes it to never call .destroy and thus never emit `close`. It might be reasonable to work around this issue in the connection timeout logic or by handling `end` events for the time being, as we work this out with an issue against Node.js (if they agree that the behavior is incorrect).

**Comment by brettkiefer** [comment]
This will be fixed when this ships on Node 10: https://github.com/nodejs/node/pull/21290. I'll leave this open until then in case anyone else sees it.

**Comment by ccs018** [comment]
Great investigation work!

**Comment by brettkiefer** [comment]
@ccs018 Thanks! The fix is out with Node 10.5.0, so I'll close this.
